File: server/php/src/Presentation/Mappers/AdditionalPaymentActionMapper.php
<?php

namespace OnlinePayments\ExampleApp\Presentation\Mappers;

use OnlinePayments\ExampleApp\Application\Domain\Enums\Common\Currency;
use OnlinePayments\ExampleApp\Application\DTOs\AdditionalPaymentActions\RequestDto;
use OnlinePayments\ExampleApp\Application\DTOs\AdditionalPaymentActions\ResponseDto;
use OnlinePayments\ExampleApp\Presentation\Models\AdditionalPaymentActions\Response;

final class AdditionalPaymentActionMapper
{
    private const SMALLEST_UNIT = 100;

    public static function mapRequest(?array $body, string $id): RequestDto
    {
        $dto = new RequestDto($id);

        $dto->amount  = isset($body['amount'])  ? (int)($body['amount'] * self::SMALLEST_UNIT)  : null;

        $dto->currency = isset($body['currency']) ? Currency::from($body['currency']) : null;

        $dto->isFinal = array_key_exists('isFinal', $body) ? (bool)$body['isFinal'] : null;

        return $dto;
    }

    public static function mapResponse(?ResponseDto $dto): Response
    {
        $response = new Response();
        if ($dto === null) {
            return $response;
        }

        $response->id = $dto->id;
        $response->status = $dto->status;
        $response->statusOutput = $dto->statusOutput;

        return $response;
    }
}

File: server/ruby/app/presentation/mappers/additional_payment_actions/additional_payment_action_mapper.rb
module Presentation
  module Mappers
    module AdditionalPaymentActions
      module AdditionalPaymentActionMapper
        SMALLEST_UNIT = 100
        module_function

        def to_dto(params, id)
          dto = Business::Dtos::AdditionalPaymentActions::RequestDto.new(id: id)

          dto.amount = params[:amount] ? (params[:amount].to_f * SMALLEST_UNIT).to_i : nil
          dto.currency = params[:currency] ? Business::Domain::Enums::Common::Currency.const_get(params[:currency]) : nil
          dto.is_final = params[:is_final]

          dto
        end

        def from_dto(dto)
          return Presentation::Models::AdditionalPaymentActions::Response.new if dto.nil?

          Presentation::Models::AdditionalPaymentActions::Response.new(
            id: dto.id,
            status: dto.status,
            status_output: dto.status_output
          )
        end
      end
    end
  end
end
